File: Euler61.hs
import Control.GroupWith
import qualified Data.Set as S
import qualified Data.Map as M
import Data.Maybe

triangle n = n * (n+1) `div` 2
square n = n * n
pentagonal n = n * (3*n-1) `div` 2
hexagonal n = n * (2*n - 1)
heptagonal n = n * (5*n - 3) `div` 2
octagonal n = n * (3 * n - 2)

triangles = map triangle [1..]
squares = map square [1..]
pentagonals = map pentagonal [1..]
hexagonals = map hexagonal [1..]
heptagonals = map heptagonal [1..]
octagonals = map octagonal [1..]

polygonals4digits =
    concat $ map only4Digits [triangles, squares, pentagonals, hexagonals, octagonals]

numDigits :: Integer -> Int
numDigits = length . show

only4Digits :: [Integer] -> [Integer]
only4Digits =
    let predicate = ((== 4) . numDigits)
    in takeWhile predicate . dropWhile (not . predicate)

first2Of4, last2Of4 :: Integer -> Integer
first2Of4 = read . take 2 . show
last2Of4 = read . drop 2 . show

foo =
    let lastDigitMap = groupWith last2Of4 polygonals4digits
        firstDigitMap = groupWith first2Of4 polygonals4digits
        pairs = [(a,b) | a <- polygonals4digits,
                         b <- fromMaybe [] (M.lookup (last2Of4 a) firstDigitMap)]
        pairsFstSet = S.fromList $ map fst pairs
        pairsSndSet = S.fromList $ map snd pairs
        -- Find pairs whose second element also has a partner
        pairsWithPartner = filter (\(_,b) -> S.member b pairsFstSet && ) pairs
    in pairsWithPartner

main = do
    print $ length foo

File: Euler61.py
import itertools

def triangle():
    i = 0
    while True:
        yield i * (i + 1) // 2  
        i += 1

def square():
    i = 0
    while True:
        yield i * i
        i += 1

def pentagonal():
    i = 0
    while True:
        yield i * (3*i - 1) // 2
        i += 1

def hexagonal():
    i = 0
    while True:
        yield i * (2 * i - 1)
        i += 1

def heptagonal():
    i = 0
    while True:
        yield i * (5 * i - 3) // 2
        i += 1

def octagonal():
    i = 0
    while True:
        yield i * (3 * i - 2)
        i += 1

def four_digits(gen):
    nums = []
    for num in gen:
        if num >= 10000:
            break
        if num >= 1000:
            nums.append(num)
    return nums

def split_digits(num):
    s = str(num)
    return (s[:2], s[2:])

triangle4 = four_digits(triangle())
square4 = four_digits(square())
pentagonal4 = four_digits(pentagonal())
hexagonal4 = four_digits(hexagonal())
heptagonal4 = four_digits(heptagonal())
octagonal4 = four_digits(octagonal())



#print(len(triangle4) * len(square4) * len(pentagonal4) * len(hexagonal4) * len(heptagonal4) * len(octagonal4))
